PE-26 | Create home page

MaxLian11 opened the pull request:
**What I did:**

- Created Home page demo with dummy records.
- I went a little beyond task PE-26: widgets Search and Category are functioning and are able to be tested using dummy records. Search bar can only search for a single keyword or better say a sequence of characters at this point.

**How to test:**

- In getQuestions function you can create different sets of dummy variables for different tests.
- Category filter must filter questions based on category, even if search was applied.
- Search bar must work properly even if category was changed in dropdown menu.

MaxLian11 committed:
PE-26 | Implemented Home Screen

MaxLian11 committed:
PE-26 | Implemented Search Bar and Category Filter

tonynguyen98 commented on lib/category_filter.dart:
```diff
@@ -0,0 +1,53 @@
+  CategoryFilter({@required void getCategoryCallback(String category)}) : _callback = getCategoryCallback;
+
+  @override
+  _MyStatefulWidgetState createState() => _MyStatefulWidgetState();
```
Can you rename `_MyStatefulWidgetState` to `_CategoryFilterState`

tonynguyen98 commented on lib/category_filter.dart:
```diff
@@ -0,0 +1,53 @@
+}
+
+/// This is the private State class that goes with MyStatefulWidget.
+class _MyStatefulWidgetState extends State<CategoryFilter> {
```
Can you rename `_MyStatefulWidgetState` to `_CategoryFilterState`

tonynguyen98 commented on lib/home_page.dart:
```diff
@@ -0,0 +1,314 @@
+
+class Home extends StatefulWidget {
+  @override
+  HomeState createState() => HomeState();
+}
+
+class HomeState extends State<Home> {
```
Not as important, but can you rename `HomeState` to `_HomeState`

MaxLian11 replied:
Sorry, forgot to change that, but we can do it later

tonynguyen98 replied:
Please fix this when you fix any merge conflicts.

tonynguyen98 commented on lib/home_page.dart:
```diff
@@ -0,0 +1,314 @@
+  String selectedCategory = "All Categories";
+
+  // Called from widget (class) category filter
+  getCategory(value) {
```
Please define the type to be a `void` here

tonynguyen98 commented on lib/home_page.dart:
```diff
@@ -0,0 +1,314 @@
+
+  // Here Eureka will get data from the database.
+  // At this point it contains just dummy lists of questions for test purposes
+  getQuestions() async {
```
Please define the type here to be `Future<void>`

MaxLian11 replied:
This caused errors on line 80: _Error: This expression has type 'void' and can't be used._
I tried `Future<dynamic>` instead and I think it worked. Will that be strongly typed?

tonynguyen98 replied:
Apologies, I didn't realize that you were returning something with this function. It looks like you're returning `List` so can you try if `Future<List>` will work? If not, then `Future<dynamic>` is fine.

tonynguyen98 commented on lib/home_page.dart:
```diff
@@ -0,0 +1,314 @@
+  // which at this point is a copy of a `data`, as no filters have been applied yet
+  @override
+  void initState() {
+    getQuestions().then((data) {
+      setState(() {
+        questionsListFiltered = questionsListFilteredSearch = questionsListFilteredCategory = data;
+      });
+    });
```
Extract this function into it's own void function and then call it here.

tonynguyen98 commented on lib/home_page.dart:
```diff
@@ -0,0 +1,314 @@
+                ? IconButton(
+                    icon: Icon(Icons.cancel),
+                    onPressed: () {
+                      // Button to drop search bar. Will keep list filtered if category filter is applied
+                      setState(() {
+                        this.isSearching = false;
+                        questionsListFilteredSearch = data; // search list is now the same as the original
+                        filterQuestionsCategory(selectedCategory); // if "All Categories", then it page is completely unfiltered
+                      });
```
extract this out to be it's own function

tonynguyen98 commented on lib/home_page.dart:
```diff
@@ -0,0 +1,314 @@
+            isSearching
+                ? IconButton(
+                    icon: Icon(Icons.cancel),
+                    onPressed: () {
+                      // Button to drop search bar. Will keep list filtered if category filter is applied
+                      setState(() {
+                        this.isSearching = false;
+                        questionsListFilteredSearch = data; // search list is now the same as the original
+                        filterQuestionsCategory(selectedCategory); // if "All Categories", then it page is completely unfiltered
+                      });
+                    },
```
extract the `onPressed` out to be it's own function

tonynguyen98 commented:
This looks really great! I have some requests about formatting changes. Ideally, we want to have each component on the screen to be it's own method. I haven't tested this yet but it looks really great!

ayeejane commented on lib/home_page.dart:
```diff
@@ -0,0 +1,314 @@
+              // Show "No results" if input text doesn't match with question title (later will be added to description too)
+              child: _buildList(questionsListFiltered, filterQuestionsCategory, getCategory)),
+
+          // Button "Create New Question" is visible if the user is in Text input mode. This feature is questionable
+          RawMaterialButton(
+            onPressed: () {},
+            fillColor: Colors.blueGrey[800],
+            splashColor: Colors.grey,
+            child: Padding(
+              padding: EdgeInsets.all(10.0),
+              child: SizedBox(
+                width: 200.0,
+                child: Text(
+                  "Create New Question",
+                  maxLines: 1,
+                  style: TextStyle(color: Colors.white),
+                  textAlign: TextAlign.center,
+                ),
+              ),
+            ),
+          ),
+        ],
+      ),
+    );
+  }
+}
```
Button for "Create New Question" should not be visible when the user is typing on the search bar. You can extract this for now or comment it out.

MaxLian11 replied:
Ok, I fixed that!
Why would you like to extract "new question" and "chat" buttons though? I think we can keep them.. They just don't have any functionalities yet

ayeejane replied:
I guess you could leave it. I just thought it could be a separate issue.

ayeejane reviewed (changes requested):
Tested it out on my end and everything works as intended. However, I think buttons for "Create New Question" and "Chat" can be extracted for now; it can be a separate issue to create those features in the near future. Also, there are a few formatting changes that Tony N. have already commented on but overall great job!

MaxLian11 committed:
PE-26 | Requested changes

txt020 reviewed (approved):
Everything looks good!!

tonynguyen98 commented on lib/home_page.dart:
```diff
@@ -0,0 +1,326 @@
+                    // Call function to apply search filter
+                    filterQuestionsSearch(value);
+                  },
+                  style: TextStyle(color: Colors.white),
+                  decoration: InputDecoration(
+                      icon: Icon(
+                        Icons.search,
+                        color: Colors.white,
+                      ),
+                      hintText: "Search Question",
+                      hintStyle: TextStyle(color: Colors.white)),
+                ),
+          centerTitle: true,
+
+          // Search icon is changed to search bar
+          actions: <Widget>[
+            isSearching
+                ? IconButton(
+                    icon: Icon(Icons.cancel),
+                    onPressed: () {
+                      searchingFalse();
+                    },
+                  )
+                : IconButton(
+                    icon: Icon(Icons.search),
+                    onPressed: () {
+                     searchingTrue();
+                    },
+                  ),
+
+            // Chat button
+            FlatButton(
+              textColor: Colors.white,
+              onPressed: () {}, // redirect to Chat page
+              child: Icon(Icons.chat_sharp, color: Colors.cyan),
+              shape: CircleBorder(side: BorderSide(color: Colors.transparent)),
+            ),
+          ],
+          backgroundColor: Colors.blueGrey[800],
+        ),
```
Extract this outside of the `Widget build` something like this: 
```dart
AppBar homeAppBar() {
return AppBar(
          // Menu button
          leading: FlatButton(
            textColor: Colors.white,
            onPressed: () {}, // open side menu
            child: Icon(Icons.menu, color: Colors.cyan),
            shape: CircleBorder(side: BorderSide(color: Colors.transparent)),
          ),

          // Title "Eureka" is changed to a search bar after search icon is clicked
          title: !isSearching
              ? Text('Eureka!')
              : TextField(
                  onChanged: (value) {
                    // Call function to apply search filter
                    filterQuestionsSearch(value);
                  },
                  style: TextStyle(color: Colors.white),
                  decoration: InputDecoration(
                      icon: Icon(
                        Icons.search,
                        color: Colors.white,
                      ),
                      hintText: "Search Question",
                      hintStyle: TextStyle(color: Colors.white)),
                ),
          centerTitle: true,

          // Search icon is changed to search bar
          actions: <Widget>[
            isSearching
                ? IconButton(
                    icon: Icon(Icons.cancel),
                    onPressed: () {
                      searchingFalse();
                    },
                  )
                : IconButton(
                    icon: Icon(Icons.search),
                    onPressed: () {
                     searchingTrue();
                    },
                  ),

            // Chat button
            FlatButton(
              textColor: Colors.white,
              onPressed: () {}, // redirect to Chat page
              child: Icon(Icons.chat_sharp, color: Colors.cyan),
              shape: CircleBorder(side: BorderSide(color: Colors.transparent)),
            ),
          ],
          backgroundColor: Colors.blueGrey[800],
        );
}

tonynguyen98 replied:
And then reference `homeAppBar()` instead. We don't want such bug chunks of code inside the `Widget build`.

tonynguyen98 commented on lib/home_page.dart:
```diff
@@ -0,0 +1,326 @@
+      ),
+
+      // List of questions and a category filter
+      body: Column(
+        children: [
+          // List of Questions
+          Expanded(
+              // Show loading circle if results are taking time
+              // Show "No results" if input text doesn't match with question title (later will be added to description too)
+              child: _buildList(questionsListFiltered, filterQuestionsCategory, getCategory)),
+
+          RawMaterialButton(
+            onPressed: () {},
+            fillColor: Colors.blueGrey[800],
+            splashColor: Colors.grey,
+            child: Padding(
+              padding: EdgeInsets.all(10.0),
+              child: SizedBox(
+                width: 200.0,
+                child: Text(
+                  "Create New Question",
+                  maxLines: 1,
+                  style: TextStyle(color: Colors.white),
+                  textAlign: TextAlign.center,
+                ),
+              ),
+            ),
+          ),
+        ],
+      ),
```
Similarly here:
```dart
Column questionsList(){
return Column(
        children: [
          // List of Questions
          Expanded(
              // Show loading circle if results are taking time
              // Show "No results" if input text doesn't match with question title (later will be added to description too)
              child: _buildList(questionsListFiltered, filterQuestionsCategory, getCategory)),

          RawMaterialButton(
            onPressed: () {},
            fillColor: Colors.blueGrey[800],
            splashColor: Colors.grey,
            child: Padding(
              padding: EdgeInsets.all(10.0),
              child: SizedBox(
                width: 200.0,
                child: Text(
                  "Create New Question",
                  maxLines: 1,
                  style: TextStyle(color: Colors.white),
                  textAlign: TextAlign.center,
                ),
              ),
            ),
          ),
        ],
      );
}
```

tonynguyen98 commented on lib/home_page.dart:
```diff
@@ -0,0 +1,326 @@
+                      ),
+                      SizedBox(height: 5),
+                      Row(children: [
+                        Text(
+                          "Asked:  ",
+                          textAlign: TextAlign.left,
+                          style: TextStyle(fontSize: 15.0, color: Colors.grey),
+                        ),
```
Here, you kinda repeat the similar code. Please make a custom Text that you can pass variables into and allow for cleaner code. Example: 
```dart
Text customText (String string, int fontSize, Colors color) {
return Text(
                          string,
                          textAlign: TextAlign.left,
                          style: TextStyle(fontSize: fontSize, color: color),
                        ),
}
```
Then you call `customText ("Asked:  ", 15.0, Colors.grey)`

tonynguyen98 reviewed (changes requested):
Some changes to provide for cleaner code.

tonynguyen98 commented:
Also, please resolve conservations when you complete them.

MaxLian11 committed:
PE-26 | More requested changes

Made AppBar and QuestionsList as separate functions. Created questionTextStyle to avoid repetitive code

MaxLian11 committed:
PE-26 | Removed "Stashed changes"

tonynguyen98 reviewed (approved):
Great work on getting your first flutter screen done! Great job!!

MaxLian11 committed:
Merge branch 'main' into PE-26-CreateHomePage

MaxLian11 committed:
PE-26 | Changes for merging to main
